File: rob599_hw3/src/median_filter.py
#!/usr/bin/env python3

import sys
import rospy

from numpy            import median, min, max, array
from sensor_msgs.msg  import LaserScan



def callback(lidar_msg):
	"""
	Callback function to filter LaserScan Data.
	"""

	if rospy.has_param('median_filter_size'):
		filterSize = rospy.get_param('median_filter_size')
	else:
		filterSize = 10

	print("\n\nFilter Size: {0}\n".format(filterSize))


	rangeSize = len(lidar_msg.ranges) -1              # Size of the lidar message


	if filterSize < 2 or filterSize > rangeSize:      # If the filter size is less than 2 or biger than the range array, then the filter is useless, just republish and exit
		pub.publish(lidar_msg)
		return


	a = round(filterSize/2)                           # Make sure the indecies remain integers



	oldRanges = array(lidar_msg.ranges)
	newRanges = []

	for i in range(0,rangeSize):                       # Cycle through the range measurmetns and apply filter

		idx = range(i-a, i+a, 1)                      # Indecies of the measurmens being filtered

		if min(idx) < 0:                              # If i is less than 1/2 the filter size, shift idx forward
			idx = idx - min(idx)

		if max(idx) > rangeSize:                      # If i is les than 1/2 the filter size from the end of the array, shift idx backward
			idx = idx - (max(idx) - rangeSize) -1

		newRanges.append( median(oldRanges[idx]) )    # Add the new filtered mesaurment to the list


	lidar_msg.ranges = newRanges                       # Reassign the range measurmetns

	pub.publish(lidar_msg)



if __name__ == '__main__':
	# Initialize the node.
	rospy.init_node('laser_filter')

	rospy.loginfo("Starting 'laser_filter' Node")

	# Set up a subscriber and publisher.
	sub = rospy.Subscriber('median_filter/laser_scan_in', LaserScan, callback )
	pub = rospy.Publisher( 'median_filter/laser_scan_filterd', LaserScan, queue_size=1 )

	rospy.spin()

File: rob599_hw3/src/median_filter.cpp
/*
	Jacob Anderson
	ROB599 HW3
	Dec 1, 2020
*/
// include the basic ROS stuff, and the Int64 definition.
#include <algorithm>
#include <vector>
#include <ros/ros.h>
#include "sensor_msgs/LaserScan.h"


// This is the callback function.  It should have a void return type, and a single argument that is a reference to a const pointer
// to the message type.  All ROS messages will allow you to use this sort of syntax.

float median(std::vector<float> &v)
{
    size_t n = v.size() / 2;
    std::nth_element(v.begin(), v.begin()+n, v.end());
    return v[n];
}



class median_filter_node
{

protected:
	ros::NodeHandle node;
	ros::Subscriber sub;
	ros::Publisher  pub;

public:

	median_filter_node(){
		// Set up the subscriber and publisher
		sub = node.subscribe("median_filter/laser_scan_in", 1, &median_filter_node::callback, this);
		pub = node.advertise<sensor_msgs::LaserScan>("median_filter/laser_scan_filterd", 1);
	}


	void callback(const sensor_msgs::LaserScan::ConstPtr& scan) {

		sensor_msgs::LaserScan newScan = *scan;																	//	New Scanline that makes use of an exisiting message

		//	Get filter size from ROS parameter or use the default of 10 if the parameter is not available
		int filterSize;

		if(node.hasParam("/median_filter_size") && node.getParam("/median_filter_size", filterSize));
		else filterSize = 10;


		int size = scan->ranges.size();																					//	Size of the laser scan

		if(filterSize < 2 || filterSize >= size) {															//	Make sure the filter size is reasonable
			pub.publish(newScan);																									//  If not, just publish the raw laserscan
			return;
			}

		ROS_INFO("Filter Size: %d",filterSize);

		// Spcaial Averaging filter
		int a = filterSize/2;
		std::vector<float> ranges;
		std::vector<float> buff;

		for (int i = 0; i < size; i++){																					//	Go through the scanline

			// Sliding window filter
			for(int j = std::max(i-a, 0); j < std::min( i+a, size ); j++ ) buff.push_back( scan->ranges[j] );

			float m = median(buff); 																							//	Get median Value of the sliding window

			buff.clear();																													// 	Clear the buffer fo rthe next pass

			ranges.push_back(m);																									//	Keep Median value
			}

		newScan.ranges = ranges;																								//	Assign the filtered values to the laser scan

		pub.publish(newScan);
	}
};


int main(int argc, char **argv) {
	// Initialize the node and set up the node handle.
	ros::init(argc, argv, "laser_filter");

	median_filter_node mfn;

	ROS_INFO("Laser Filter Node Started");
	// Give control over to ROS.
	ros::spin();

	return 0;
}
